src/utf8.cpp

```diff
@@ -0,0 +1,346 @@
+        // This is a no-op if the case is already correct
+        const utf8proc_property_t *originalCodepointProperty = utf8proc_get_property((int32_t)codepoint);
+        if (originalCodepointProperty->casefold_seqindex != UINT16_MAX) {
+            if (originalCodepointProperty->category & (UTF8PROC_CATEGORY_LU | UTF8PROC_CATEGORY_LL)) {
+                bool originalCPWasUpper = originalCodepointProperty->category == UTF8PROC_CATEGORY_LU;
+                for (size_t wbIndex = 0; wbIndex < decomposedLength; ++wbIndex) {
+                    if (originalCPWasUpper) {
+                        wbBuffer[wbIndex] = utf8proc_toupper(wbBuffer[wbIndex]);
+                    } else {
+                        wbBuffer[wbIndex] = utf8proc_tolower(wbBuffer[wbIndex]);
+                    }
+                }
+            }
+        } else {
+            // We're forcing a case conversion back and forth if necessary to catch the few stragglers like ı (i)
+            for (size_t wbIndex = 0; wbIndex < decomposedLength; ++wbIndex) {
+                const utf8proc_property_t *codepointProperty = utf8proc_get_property(wbBuffer[wbIndex]);
+
+                // If this is a uppercase codepoint, we do a tolower then toupper
+                if (codepointProperty->category == UTF8PROC_CATEGORY_LU) {
+                    wbBuffer[wbIndex] = utf8proc_toupper(utf8proc_tolower(wbBuffer[wbIndex]));
+                }
+
+                // If this is a lowercase codepoint, we do a toupper then tolower
+                else if (codepointProperty->category == UTF8PROC_CATEGORY_LL) {
+                    wbBuffer[wbIndex] = utf8proc_tolower(utf8proc_toupper(wbBuffer[wbIndex]));
+                }
+            }
+        }
+    }
+
+    return decomposedLength;
+}
+
+bool normalize_string(char **_utf8Buffer, uint64_t &bufferLength)
+{
+    const char *utf8Buffer = *_utf8Buffer;
+    int32_t defaultInFlightBuffer[128], *inFlightBuffer = defaultInFlightBuffer;
+    size_t inFlightBufferLength = 128;
+    std::vector<ScratchpadChunck> scratchPad;
+
+    // A tricky part of this conversion is that the output size is totally unknown, but we want to be efficient with our
+    // allocations. We're going to write the glyph we're normalising in a static buffer (if possible) and write the
+    // filtered, normalized results in a bunch of semi-fixed buffer (the scratchpad) Only when the conversion is over
+    // will we allocate the final buffer and copy everything in there.
+    scratchPad.reserve(8);
+    scratchPad.emplace_back(bufferLength > 1024 ? bufferLength : 1024);
+
+    uint32_t codepoint;
+    uint64_t position = 0;
+    while ((codepoint = fetch_next_codepoint(utf8Buffer, position, bufferLength)) != UTF8_EOF) {
+        // Ignore invalid glyphs
+        if (codepoint == UTF8_INVALID) {
+            continue;
+        }
+
+        size_t decomposedLength = normalize_codepoint(codepoint, inFlightBuffer, inFlightBufferLength);
```

Comment:
To avoid codepoint explosion (as you called it), perhaps we should limit the size of the `inFlightBuffer` as it could be used by an attacker to force the WAF to run out of memory.